File: [0098][Medium][Validate Binary Search Tree]/Validate_Binary_Search_Tree_2.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
// leetcode time     cost : 1 ms
// leetcode memory   cost : 38.7 MB
class Solution {
    long pre = Long.MIN_VALUE;
    boolean flag = true;
    public boolean isValidBST(TreeNode root) {
        inOrder(root);
        return flag;
    }
    public void inOrder(TreeNode root)
    {
        if(root == null)
            return;
        inOrder(root.left);
        if(pre >= root.val){
            flag = false;
            return;
        }
        pre = root.val;
        inOrder(root.right);
    }    
}

File: [0098][Medium][Validate Binary Search Tree]/Validate_Binary_Search_Tree_2.cpp
#include <stdlib.h>
using namespace std;
/*   Definition for a binary tree node. */
struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode(int x) : val(x), left(NULL), right(NULL) {}

};
// leetcode time     cost : 32 ms
// leetcode memory   cost : 20.8 MB
//C++ inorder recursion:
class Solution {
public:
    int* last = NULL;
    bool isValidBST(TreeNode* root) {
        if (root){
            if(!isValidBST(root->left)) 
                return false;
            if (last && *last>=root->val) 
                return false;
            last = &root->val;
            if(!isValidBST(root->right)) 
                return false;
            return true;
        }
        else 
            return true;
    };
};